README.md
# biotope-build
Build Package for biotope-boilerplate.

## Features
- Typescript + Babel
- ES6 unglify
- SCSS
- Autoprefixer
- Icon Font
- SVG Sprite generation
- HBS for templating
- ESLint
- Dev Servers

## Options & Settings

All biotope-build options can be set in local /projectConfig.js (<5.x /projectConfig.json)

See default config: <https://github.com/biotope/biotope-build/blob/master/config.js>

### framework internals
#### global settings
```
{
   <String> src: src, //base folder for all the project files, defaults to 'src'
   <String> dev: dev, //temporary folder for all files that get created in the serve/build process, defaults to '.tmp'
   <String> dist: dist, //build folder for the finalized files when running a build process, defaults to 'dist'
   <String> node: node, //node_modules base folder, defaults to 'node_modules'
   <String> cwd: cwd, //process.cwd(), current working dir of node project ###TODO should maybe be private###
   <boolean> isWin: isWin, //checks the os and sets the variable to true if running on a windows system
   <boolean> debug: false, //toggle debbuging prompts from the tasks (static:hb, browserConfig, partialHelper), default is false
   <String> dataObject: 'data' //name of the global data object, which can be used in static handlebars templates via {{data.key}}
   <String> resources: '/resources', //resource folder inside the src folder, needed to run multiple tasks
   <String> components: '/components' //component folder corresponding to the resource folder
   <String> handlebarsHelper: '/js/handlebars.helper.js', //path to dynamic handlebar helpers relative to the resources folder
   <Object> tasks: {}, //see tasks
   <Object> externalResources: {} //see external resources
   <Object> bowerResources: {} //see bower resources
}
```
#### tasks settings
You can toggle certain task in biotope-build. Every Task can be configurated with a boolean.

This list shows the configurable tasks as well as a brief description about what they do.
```
{
	browserSupport: true, //creates an overview of the support browser from a *.json file
	cleanCss: true, //minify CSS files for build
	cssStats: true, //checks the maximum selectors in each css file, fails if there are more than 4096 (IE 9 issue)
	favicons: true, //creates and embeds the favicons as well as the mobile icons from an image, default image root is '/resources/'
	handlebars: true, //creates dynamic handlebar templates from hbs files, static page generation is handled in a seperate task
	iconfont: true, //creates the iconfont from svg images, default icon path is '/resources/icons'
	image: true, //compress images stores in '/resources/img'
	linting: true, //linting of js/ts/scss files
	sass: true, //disables the sass compiler, disabled if you use vanilla css
	uglify: true, //uglifies js files in the build
	webpack: true //compiles typescript and transpiles ES6 Code in *.ts files
}
```
#### externalResources
If you use external resources for the final build you can add those with this object. The module name references the module inside the node_modules folder. The file name needs to be relative to the module.

JS files then get copied to '.tmp/resources/js/vendor/', CSS files to '.tmp/resources/css/vendor/'. When using scss files, reference the node_modules folder directly from there.
```
{
	'moduleName': 'singleFile.ext',
	'otherModuleName': ['firstFile.ext', 'folder/secondFile.ext2'],
}
```

### plugin defaults
#### autoprefixer
Autoprefixer uses the .browserlistrc file from your projects root, or otherwise default options
Repo: https://github.com/postcss/autoprefixer#options


#### browserSupport
Set the file path to the corresponding *.json file to create a browser support overview.

```
browserSupport: {
	file: cwd + '/browserSupport.json'
},
```

#### checkDependencies
Check dependencies only uses default options.
Repo: https://github.com/mgol/check-dependencies

#### cleanCss
Clean CSS uses default options with rebase set to false to prevent rewriting file paths.
Repo: https://www.npmjs.com/package/gulp-clean-css
Repo: https://github.com/jakubpawlowicz/clean-css
```
cleanCss: {
	rebase: false
},
```

#### cssStats
cssStats exits on error in order to break the build process.
Repo: https://github.com/cssstats/gulp-cssstats

```
cssstats: {
	exit: true
},
```

#### connect
Set the port and the folders being watch for connect/livereload.
Repo: https://github.com/avevlad/gulp-connect

```
connect: {
		port: 9000,
		globs: [
			dev + '/**/*.*',
			src + '/resources/js/**/*.js',
			src + '/resources/bower_components/**/*',
			src + '/_mock/**/*',
			src + '/_config/**/*',
			'!' + dev + '/_mock/**/*',
			'!' + dev + '/_config/**/*',
			'!' + dev + '/_assets/**/*',
			'!' + dev + '/resources/js/vendor/**/*.js',
			'!' + dev + '/resources/css/**/*.map',
			'!' + dev + '/resources/bower_components/**/*',
			'!' + dev + '/resources/js/handlebars.templates.js'
		]
	},
```

#### favicons
This creates all the favicons and a manifest.json from an image file.
Repo: https://github.com/evilebottnawi/favicons

```
favicons: {
	appName: "gulp-frontend-boilerplate",
	background: "#020307",
	path: "favicons/",
	display: "standalone",
	orientation: "portrait",
	version: 1.0,
	logging: false,
	online: false,
	html: "htmlhead.favicons.html",
	pipeHTML: true,
	replace: true,
	icons: {
		appleStartup: false
	}
},
```

#### handlebars
Dynamic handlebar templates and partials get created and added to a namespace.
Repo: https://github.com/lazd/gulp-handlebars

```
handlebars: {
	templateWrap: 'Handlebars.template(<%= contents %>)',
	partialWrap: 'Handlebars.registerPartial(<%= processPartialName(file.relative) %>, Handlebars.template(<%= contents %>));',
	namespace: 'ffglobal.configuration.data.tpl',
	noRedeclare: true
},
```

#### iconfont
Icon font settings are split into gulp-svgicons2svgfont settings and the path settings.
Repo: https://github.com/nfroidure/gulp-svgicons2svgfont

```
iconfont: {
	fontName: 'Icons',
	prependUnicode: true,
	timestamp: Math.round(Date.now() / 1000),
	normalize: true
},

iconfontCss: [
	{
		fontName: 'Icons',
		path: src + '/resources/scss/fonts/iconfont/_icons.scss',
		targetPath: '../../../../.iconfont/_icons.scss',
		fontPath: '../fonts/icons/',
		cssClass: 'icon'
	}, {
		fontName: 'Icons',
		path: src + '/resources/scss/fonts/iconfont/_iconStyles.scss',
		targetPath: '../../../../.iconfont/_iconStyles.scss',
		fontPath: '../fonts/icons/',
		cssClass: 'icon'
	}
],
```

#### image
Minifiy the images inside "/resources/img".
Repo: https://github.com/sindresorhus/gulp-imagemin

```
image: {
	verbose: true
},
```

#### lec
Force line-endings to 'LF' formatting to create unification across platform.
Repo: https://github.com/iShafayet/gulp-line-ending-corrector

```
lec: {
	verbose: false,
	eolc: 'LF',
	encoding: 'utf8'
},
```

#### livereload
Specifies the port for using livereload. Runs with connect. See connect settings.
```
livereload: {
	port: 35729
},
```

#### modernizr
Runs modernizr tests and creates polyfills.
Repo: https://github.com/doctyper/gulp-modernizr
Options: https://github.com/Modernizr/customizr#config-file
```
modernizr: {
	options: [
		"setClasses",
		"addTest"
	],
	excludeTests: [
		'hidden'
	]
},
```

#### sass
Sass compiler based on node-sass.
Repo: https://github.com/dlmanning/gulp-sass
Options: https://github.com/sass/node-sass#options
```
sass: {
	includePaths: []
},
```

#### sassLint
Linting of sass files. Default options. Rules need to be defined.
Repo: https://github.com/sasstools/gulp-sass-lint
```
sassLint: {},
```

#### ugflify
Javascript uglification settings.
Repo: https://github.com/terinjokes/gulp-uglify

```
uglify: {
	preserveComments: 'license',
	sourcemaps: false,
	folders: ['js', 'ts'],
	ignoreList: []
},
```

#### watch
Watch task uses polling on windows system, therefore the interval is increased to reduce cpu usage.
Repo: https://github.com/floatdrop/gulp-watch

```
watch: {
	usePolling: isWin,
	interval: (isWin ? 250 : 100)
},
```

#### webpack
Add files to an ignoreList, set path relative to src path. Webpack settings are set via webpack.config.js
Repo: https://github.com/webpack/webpack
```
webpack: {
	ignoreList: []
}
```

## others
There are other tasks in use, but they can not be modified via a projectConfig.js file. If this is needed please open an issue or send a pull request.

## Local development
1. Clone [biotope-boilerplate repository](https://github.com/biotope/biotope-boilerplate)
2. Install locally with `npm`
3. Clone biotope-build repository
4. Install dependencies in biotope-buid with `npm`
5. Run `npm link` in @biotope/build repo
6. Run `npm link @biotope/build` in biotope-boilerplate repo
7. Happy developing 👋

### Undo local dev changes
1. Run `npm unlink` in @biotope/build repo
2. Run `npm unlink @biotope/build` in biotope-boilerplate repo
3. Run `npm` in biotope-boilerplate repo

## Troubleshooting, Bugs & Issues
### Knows issues
- Missing documentation (on the roadmap)
- Poor performance (on the roadmap)

If you encounter errors, please submit an issue.


CHANGELOG.md
# Changelog

<details>
  <summary>
	Placeholder section to be used for future releases. Make no changes here.
  </summary>

## [Unreleased] - YYYY-MM-DD
### Added
- for new features.
### Changed
- for changes in existing functionality.
### Deprecated
- for soon-to-be removed features.
### Removed
- for now removed features.
### Fixed
- for any bug fixes.

</details>

## [7.1.0] - 2019-08-06
### Added
- Make weback.config.js extendable

### Changed
- remove splitChunks from default webpack.config.js because of breaking none bio-element projects

## [7.0.0] - 2019-07-18
### Added
- Added enhanced assets copying (assets in components folders are copied now to dist)

### Changed
- BREAKING: you have to add a .browserlistrc file to your projects root folder you can use the default: https://github.com/biotope/biotope-boilerplate/blob/master/.browserslistrc. Remove "browsers" option from autoprefixer section in projectConfig.js if it exists
- BREAKING: typescript updated from version 2.9.2 to 3.5.3 please see breaking changes in typescript for migration https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes
- BREAKING: All linters are deactivated by default now and will be removed in 8.0 (Please see Deprecation section for more information). If you need to, You can still activate them by overwriting defaults in your projectConfig.js, otherwise if you have them activated in your projectConfig.js your should remove the following tasks: cssStats, htmlhint, linting
- Most dependencies updated
- Stability fixes
- Performance improvements

### Deprecated
- All linters built into biotope-build are now deprecated. Please switch to the biotope-quality-gate instead: https://github.com/biotope/biotope-quality-gate

### Removed
- BREAKING: removed the support for Bower package manager

## [6.2.0] - 2019-06-06
### Added
- Added file loader for videos and images to webpack
- Added cors for localhost server

## [6.0.0] - 2019-02-27
### Added
- Add svg loader for scss and ts files
### Changed
- Exchanged awesome-typescript-loader with ts-loader for a more common loader
- Update eslint config to ignore src/resources/js/polyfills
### Fixed
- fix scss task to build component scss files again

## [5.6.0] - 2019-02-26
### Changed
- Upgrade Babel
- Update eslint config to ignore src/resources/js/polyfills
### Removed
- Remove check dependencies due to failure of custom tags

## [5.5.0] - 2018-11-23
### Added
- Add configurable ts entry points.

## [5.4.5] - 2018-11-16
### Changed
- updated readme
### Removed
- removed eot font creation from iconfont task
### Fixed
- sass compilation sometime not triggering when updateding partials
- css minification not working
- css charset getting removed

## [5.4.4] - 2018-10-26
### Fixed
- fixed changing sass partials not triggering recompile
### Changed
- added configuration to webpack to enable sass entry points - default is false

## [5.4.3] - 2018-10-19
### Fixed
- changing a sass files triggers compilation of all sass files -> now only compiles the changed file

## [5.4.2] - 2018-09-17
### Added
- Merge env data into handlebars global data object under `data.env`
- Add support for `cross-env`

## [5.4.1] - 2018-08-30
### Fixed
- ES6 modules from node_modules are now transpiled by Webpack (https://github.com/biotope/biotope-build/issues/129)

## [5.4.0] - 2018-08-17
### Changed
- BREAKING: After updateing to 5.4.0 rename resources/scss/fonts/iconfont/_icons.scss → src/resources/scss/fonts/iconfont/_icons.tpl
### Added
- Support for env files added
- support for svg sprites added
- Support for ES6 added (uglifyjs)

## [5.3.9] - 2018-06-27
### Updated
- Updated all node modules to newest versions

## [5.3.8] - 2018-06-21
### Fixed
- Disabled Uglify task prevents components from being copied (https://github.com/biotope/biotope-build/issues/73)
### Added
- Welcome message in biotope console output :sparkles:
- Added error message log for missing local projectConfig.js   

## [5.3.6] - 2018-06-05
### Fixed
- Webpack tasks is broken when no TS files are available

## [5.3.5] - 2018-05-30
### Fixed
- Heap out of memory bug (https://github.com/biotope/biotope-build/issues/48)
### Changed
- Local tsconfig.json needs update:
```
	"awesomeTypescriptLoaderOptions": {
		"useCache": true,
		"reportFiles": [
			"src/**/*.{ts,tsx}"
		]
	},
```

## [5.3.4] - 2018-05-20
### Removed
- Iconfont path checker

## [5.3.3] - 2018-05-20
### Changed
- Webpack4 Update - also added entry points support instead of piping vinyl files to webpack
### Fixed
- Useref task now handles html and hbs extension
- Useref task now uses bioHelpers   
### Added
- Iconfont path checker

## [5.3.2] - 2018-04-24
### Fixed
- Webpack now loads JS and JSX files with babel loader

## [5.3.1] - 2018-04-18
### Fixed
- Missing semicolon in biotope handlebars helper injection

## [5.3.0] - 2018-04-17
### Changed
- Complete Handlebars Task Refactoring
- BREAKING: Change `handlebars.helpers.js` according to `https://github.com/biotope/biotope/blob/5.3.0/src/resources/js/handlebars.helper.js`
- BREAKING: Change `index.hbs` according to `https://github.com/biotope/biotope/blob/5.3.0/src/index.hbs`
- BREAKING: Change `browserSupport.hbs` according to `https://github.com/biotope/biotope/blob/5.3.0/src/browserSupport.hbs`
- BREAKING: Rename all {{ frontMatter.* }} expressions to {{ data.frontMatter.* }} - mostly used in layout template as title tag

## [5.2.1] - 2018-03-14
### Added
- Copies "_mock" folders inside components to dist during build
- Writes version specified in package.json to resources/VERSION

## [5.2.0] - 2018-02-25
### Changed
 - BREAKING: Repository name changed to github.com/biotope/biotope-build. Adjust your package.json
 - BREAKING: Prefixed all handlebars helper functions with "bio", for example the helper "include" is now named "bioInclude". This reduces chance of naming collision with variables [(see)](https://github.com/biotope/biotope-build/pull/45/files)

config.js
const cwd = process.cwd();
const src = 'src';
const dev = '.tmp';
const dist = 'dist';
const node = 'node_modules';

const path = require('path');
const os = require('os');
const isWin = /^win/.test(os.platform());

module.exports = {
  global: {
    src: src,
    dev: dev,
    dist: dist,
    node: node,
    cwd: cwd,
    isWin: isWin,
    debug: false,
    dataObject: 'data',
    resources: '/resources',
    components: '/components',
    handlebarsHelper: '/js/handlebars.helper.js',
    tasks: {
      browserSupport: true,
      cleanCss: true,
      cssStats: false,
      favicons: true,
      handlebars: true,
      htmlhint: false,
      iconfont: true,
      svgSprite: true,
      image: true,
      linting: false,
      sass: true,
      uglify: true,
      version: true,
      webpack: true
    },
    tsEntryPoints: [
      'resources/ts/**/*.ts',
      'components/**/*.ts'
    ],
    externalResources: {},
    bowerResources: {}
  },

  //=== Plugins ===//
  autoprefixer: {
    grid: true
  },

  browserSupport: {
    file: path.join(cwd, 'browserSupport.json'),
    property: 'data.browserSupport'
  },

  checkDependencies: {},

  cleanCss: {
    rebase: false
  },

  connect: {
    host: '0.0.0.0',
    port: 9000,
    historyFallbackIndex: '/index.html',
    globs: [
      dev + '/**/*.*',
      src + '/resources/js/**/*.js',
      src + '/_mock/**/*',
      src + '/_config/**/*',
      '!' + dev + '/_mock/**/*',
      '!' + dev + '/_config/**/*',
      '!' + dev + '/_assets/**/*',
      '!' + dev + '/resources/js/vendor/**/*.js',
      '!' + dev + '/resources/css/**/*.map',
      '!' + dev + '/resources/js/handlebars.templates.js'
    ]
  },

  cssstats: {
    exit: true
  },

  env: {
    files: {
      production: 'prodvars.env',
      staging: 'stagingvars.env',
      development: 'devvars.env'
    }
  },

  favicons: {
    appName: 'biotope',
    background: '#020307',
    path: 'favicons/',
    display: 'standalone',
    orientation: 'portrait',
    version: 1.0,
    logging: false,
    online: false,
    html: 'htmlhead.favicons.html',
    pipeHTML: true,
    replace: true,
    icons: {
      appleStartup: false
    }
  },

  handlebars: {
    templateWrap: 'Handlebars.template(<%= contents %>)',
    partialWrap: 'Handlebars.registerPartial(<%= processPartialName(file.relative) %>, Handlebars.template(<%= contents %>));',
    namespace: 'biotope.configuration.data.tpl',
    noRedeclare: true
  },

  frontMatter: {
    property: 'frontMatter',
    remove: true
  },

  iconfont: {
    fontName: 'Icons',
    prependUnicode: true,
    timestamp: Math.round(Date.now() / 1000),
    normalize: true
  },

  iconfontCss: [
    {
      fontName: 'Icons',
      path: src + '/resources/scss/fonts/iconfont/_icons.tpl',
      targetPath: '../../../../.iconfont/_icons.scss',
      fontPath: '../fonts/icons/',
      cssClass: 'icon'
    }, {
      fontName: 'Icons',
      path: src + '/resources/scss/fonts/iconfont/_iconStyles.scss',
      targetPath: '../../../../.iconfont/_iconStyles.scss',
      fontPath: '../fonts/icons/',
      cssClass: 'icon'
    }
  ],

  image: {
    verbose: true
  },

  lec: {
    verbose: false,
    eolc: 'LF',
    encoding: 'utf8'
  },

  livereload: {
    port: 35729
  },

  modernizr: {
    options: [
      'setClasses',
      'addTest'
    ],
    excludeTests: [
      'hidden'
    ]
  },

  sass: {
    includePaths: ['node_modules']
  },

  sassLint: {},

  uglify: {
    options: {
      output: {
        comments: /@preserve|@license|@cc_on/i
      }
    },
    sourcemaps: true,
    folders: ['js', 'ts'],
    ignoreList: []
  },

  watch: {
    usePolling: isWin,
    interval: (isWin ? 250 : 100)
  },

  webpack: {
    ignoreList: ['/**/*.spec.ts'],
    watchScss: false,
    tsWatchPatterns: [
      'resources/ts/**/*.ts',
      'components/**/*.ts'
    ]
  }
};

try {
  const projectConfig = require(path.join(cwd, 'projectConfig.js'));
  if (projectConfig) {
    const merge = require('lodash.merge');
    merge(module.exports, projectConfig);
  }
} catch (e) {
  const colors = require('colors/safe');
  console.log(colors.red('Error: ', e));
}
